Add image pull progress KEP

saschagrunert opened the pull request:
**State: work in progress / in review**

Related discussions: 
- https://github.com/kubernetes/kubernetes/issues/19077
- https://groups.google.com/forum/#!topic/kubernetes-sig-node/JHEus_TlZzA

Feedback and input is highly appreciated.

Open Discussion Points:
- https://github.com/kubernetes/enhancements/pull/1338#discussion_r342623888 (and following)
- https://github.com/kubernetes/enhancements/pull/1338#discussion_r344498458 (and following)

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,314 @@
+and `FailedToPullImage` will be re-used and extended with the additional
+metadata provided by the CRI. This means that the container will still reside in
+`ContainerStateWaiting` during the image pull, but the events regarding the
+image pull process will be thrown periodically.
```
This has non-trivial scalability implications. I would like to have a better analysis at the KEP level.

In fact, that is a great candidate for pilot of scalability-related part of PRR:
https://github.com/kubernetes/community/pull/4184

@johnbelamaric - FYI

saschagrunert replied:
I agree. I thought about having the same update period as the CRI stream updates are coming in. What about making it configurable? I think a value around 1-30 seconds would be reasonable depending on the bandwidth.

saschagrunert replied:
Alright I will attach the answers in the first comment.

saschagrunert replied:
I tried to answer all the questions from the questionnaire (see PR body), PTAL and bear with me If I misinterpreted something. :see_no_evil:

wojtek-t replied:
I don't see any of those answered anywhere - can you please point me to them?
[BTW - it's a crazy week for me, so I may not get to it before end of week...]

saschagrunert replied:
Yes, the first comment on this PR contains a drop down menu where the answered are listed :)

![Peek 2019-10-28 12-32](https://user-images.githubusercontent.com/695473/67675324-2517f280-f97f-11e9-86d0-72f89f627b53.gif)

wojtek-t replied:
Aaah I see - now. Thanks. Will take a look.

dims replied:
LOL @saschagrunert :)

saschagrunert commented:
Ready for a first review. :pray:

saschagrunert commented:
> @saschagrunert - I'm sorry, I didn't get to it yet - crazy week. Will get to this as one of the first things on Monday.

Alright, thanks for letting me know :pray:

wojtek-t commented:
@sakshamsharma - we were discussing the topic where the production readiness questionaire should be places, and the current thinking is that it should actually be part of the KEP. So if you could move it from the PR description to the KEP itself, it would be great.

Below couple comments to your answers

>>   - **Will enabling / using the feature result in any new API calls? ...
   
>    Yes, we throw new events during the image pull process. The amount can be
    configured on the kubelet as well. The data for the events are coming from
    the corresponding container runtime, which streams the result to the
    kubelet in the same amount. There is no further state involved than the
    events one.

What I would expect in the answer for this question is also some number(s). Obviously I agree that it's super hard to provide one exact number in this case, but we can try doing some estimations.
As an example, sth like that: assuming that an average image is downloading 1 minute (feel free to modify the number), then with 100pods/s throughput it would give us 6000 pods for which an image is being downloaded at a given time. So depending on the configuration (1s-30s notifications), it would result in 200-6000 QPSes to apiserver.
[I'm not clayiming the numbers are good - I'm more trying to show a pattern.]


>>   - **Will enabling / using the feature result in supporting new API types? How
    many objects of that type will be supported (and how that translates to
    limitations for users)?**

>    Yes, the pull related events `PullingImage`, `PulledImage` and
    `FailedToPullImage` will contain additional metadata, like the pull
    progress.

That's not actually a new API type - you are suggesting using the existing Event API type.
So the answer for this question is actually "no" (you will just be setting appropriate fields to new values).

>>   - **Will enabling / using the feature result in increasing size or count of
    the existing API objects?**

>    Yes, the amount of events increase.

Can you please try to estimate the number e.g. in worst case (e.g. something that I described above for qps)?

>>   - **Does this feature depend on any specific services running in the cluster (e.g., a metrics service)?**

>    Yes the container runtime needs to support the new CRI API.

Does the API already exist? If not, is there already a plan on creating it?
[BTW - it would be useful to name the method field here.]

>>   - **How does this feature respond to complete failures of the services on which it depends?**

>    (I think) a connection to the runtime is not possible at all, but this
    needs further clarification.

I guess the point of this question is mainly to:
(a) understand if your feature will work
(b) if it won't break other features (e.g. pod startup)
I hope that the answer will be "we won't be reporting status, but it shouldn't break anything else".


>>   - **What are the service level indicators an operator can use to determine the health of the service?**

>    The runtime and kubelet health status information.

I disagree with this answer. In my current cluster runtime and kubelet are healthy, and this feature doesn't work.
I think what we should do is sth like: "kubelet is reporting a prometheus metric for number of failred those new calls to CRI and number of events it failed to report. Some aggregation of both across all nodes is the SLI".



I will have couple more comments, but let's start with this one.
[BTW - thanks a lot for helping us piloting the questionaire - it's extremely useful.]

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,231 @@
+metadata provided by the CRI. This means that the container will still reside in
+`ContainerStateWaiting` during the image pull, but the events regarding the
+image pull process will be thrown periodically. The amount of how often the
+kubelet throws this event should be configurable in a range between 1s and 1min.
```
BTW - I'm not 100% convinced about the use of Events type for it. 
I would like to see a bit more thoughts about what will it mean once we will migrate everywhere to new events API:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/events-redesign.md
https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190131-event-series.md

[Note that we already migrated scheduler in 1.16, and we're planning to do the remaining components, including kubelet in 1.18]

saschagrunert replied:
Sounds reasonable, maybe @kubernetes/sig-node-api-reviews could provide more input here?

dixudx replied:
> the events regarding the image pull process will be thrown periodically

What's the period time? 30s? 1min?
Or the progress for every 10% will be reported?

This will definitely become a heavy burden when the cluster scales out. Thinking about deploying a DaemonSet.

saschagrunert replied:
> So to make my point a bit more explicit - I would like to see some discussion whether we should use Events or some dedicated CRD for this maybe. Both has advantages and disadvantages and the answer isn't clear for me.

Can you please point out the advantages of using a CRD over an Event in this case? This is not clear to me in all shapes.

> What's the period time? 30s? 1min?

The period is the configured time. :)

> Or the progress for every 10% will be reported?

Hm I don't think that this will increase the usability for the user if a large image only reports at every 10%. I think a time-constraint makes most sense, WDYT?

> This will definitely become a heavy burden when the cluster scales out. Thinking about deploying a DaemonSet.

Do you think that we should add a global limit per image/node?

wojtek-t replied:
> Can you please point out the advantages of using a CRD over an Event in this case? This is not clear to me in all shapes.

Event is already defined API.
- Do we really want to change hide the number in the event message?
- Events have TTL=1h - do we really want it? Maybe CR with OwnerReferences set to the pod is better?
- Events are being merged (though we need to check based on what). What exactly do we want? Having a single CR per pod that is being explictly updated (rather than `emited` as events are)?
- If we use a dedicated type, I can imagine dedicated client, with explicit QPS limits. Medium/long term, a dedicated bucket for request priority/fairness quota.

Just some advantages that I can see. but there are also disadvantages (like dependency on the new CRD, how this is going to be installed, additional maintainence burden, ...).

derekwaynecarr replied:
i think we need to figure out a budget on the number of writes back to an api server we anticipate during pod startup.  i see this as potentially overwhelming the api server in dense clusters.

saschagrunert replied:
Yeah I agree. So after the KubeCon and some discussions around the topic API vs CRD I tend to go the more future proof CRD way. Would it be okay to just state:

> The image pull progress feature will be deployed via a Custom Resource Definition (CRD),
> which will be deployed automatically via a new feature gate `ImagePullProgress`.

WDYT? If we can agree on that I'm going to update the PR.

pohly replied:
So is there now support for deploying a CRD that core Kubernetes can depend on? Do you have a pointer to an example or documentation for how this "will be deployed automatically via a new feature gate" part can be implemented? I'm curious because I am facing the same question in another KEP. Thanks!

wojtek-t replied:
> So is there now support for deploying a CRD that core Kubernetes can depend on?

I think that I don't fully agree on the fact that it will "depend on it". Basically, this is more information surfaced for debugging (the fact that we consider putting it in events, which are not guaranteed to be delivered, is also a good sign of it). So I guess what I'm saying here, is that it's not a critical feature - it's more "you will have less information for debugging why things are taking so long".

saschagrunert replied:
@wojtek-t do you think that we can ship the CRD with the kubelet (feature gate)? Or do you think it should live outside of the k/k tree?

wojtek-t replied:
I don't know - we should discuss this topic on SIG Architecture. Next meeting will happen next week - we should probably add this topic to agenda.

saschagrunert replied:
That would be wonderful, is the meeting somehow timezone friendly to people from Europe? :)

wojtek-t replied:
11am PST - so not perfect, but attendable

pohly replied:
@wojtek-t :
> I think that I don't fully agree on the fact that it will "depend on it".

Let's not rathole on how important the CRD is for Kubernetes.

The key question is whether any code in the kubernetes/kubernetes repo will depend on the client bindings for the CRD (because that affects where to maintain, generate and version them) and who's going to install the CRD when bringing up a cluster. If that part is left to tools like kubeadm, then Kubernetes itself has to work also when the CRD is not installed.

@saschagrunert:
> That would be wonderful, is the meeting somehow timezone friendly to people from Europe? :)

https://github.com/kubernetes/community/tree/master/sig-architecture

Okayish I guess, but I know what you mean... :grimacing:

saschagrunert replied:
Okay cool, I will attend :)

saschagrunert commented:
> [BTW - thanks a lot for helping us piloting the questionaire - it's extremely useful.]

Thank you for the valuable input, I changed the mentioned parts of the questionnaire and moved it into the KEP.

wojtek-t commented:
>> [BTW - thanks a lot for helping us piloting the questionaire - it's extremely useful.]

> Thank you for the valuable input, I changed the mentioned parts of the questionnaire and moved it into the KEP.

@saschagrunert Thanks a lot - I will take another pass tomorrow.
BTW - given the questionaire is in the pilot phase, if you have any suggestions how to make questions more understandable (or if any changes/tweaks to them would let you better understand the questions (read to match what my personal expectations were)), I would be more than happy to hear them. The goal is to have all people understand my intention not me thinking that the questions are good :)

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+
+- **How can an operator determine if the feature is in use by workloads?**
+
+  Checking if its enabled in the kubelet configuration.
```
I would say that this is N/A - the feature isn't per-se used by workloads, it's only used by system-components (Kubelet in particular).

saschagrunert replied:
I agree and adapt as suggested.

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+
+- **What are reasonable service level objectives for the feature?**
+
+  (I'm unsure about this question)
```
I would say N/A - this feature is more to make debugging easier. I wouldn't anticipate the SLO for it.

saschagrunert replied:
Alright. We may update the questions where a `N/A` is possible and indicate that this answer is fine, too :)

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+
+- **How can those be detected via metrics or logs?**
+
+  Yes, the kubelet will log the failure.
```
Any metric you're planning to add?

saschagrunert replied:
I would add the metric to the container runtime. I don't think that the kubelet needs an additional metric for it, WDYT?

saschagrunert replied:
Hm in terms of troubleshooting, I think an additional kubelet metric would make sense...

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+- **What steps should be taken if SLOs are not being met to determine the
+  problem?**
+
+  Open up an issue in kubernetes/kubernetes or contacting SIG node.
```
No SLO is anticipated.

saschagrunert replied:
Do you think we can provide the responder to this questionnaire more insights what an SLO is? This might be an obstacle for some (I had to google for a definition honestly).

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+  - **periodic API calls to reconcile state (e.g. periodic fetching state,
+    heartbeats, leader election, etc.)**
+
+  Yes, we throw new events during the image pull process. The amount can be
```
s/amount/frequency/ would probably be better word

saschagrunert replied:
Fixed :+1:

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+    heartbeats, leader election, etc.)**
+
+  Yes, we throw new events during the image pull process. The amount can be
+  configured on the kubelet as well. The data for the events are coming from
```
I think that the rest of this paragraph is not relevant to this question, so we may probably remove it.

saschagrunert replied:
Alright I removed everything from "The data for the events..."

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+  the corresponding container runtime, which streams the result to the
+  kubelet in the same amount. There is no further state involved than the
+  events one.
+
```
I would mention worst case - if downloading is extremely slow, we can (in theory) be sending event per pod every X seconds (where X is what you configure).

saschagrunert replied:
Okay, I'll add a note about this.

wojtek-t commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+  the image pull, which would result in at least 200 QPS to the API Server.
+
+  If we configure the feature to an event period of 1 second, then it would
+  result in 800 QPS to the API Server.
```
I wouldn't give single number - instead I would say that depending on the interval it will be between 0 and 800 additinal QPS in this scenario.

saschagrunert replied:
Yes I agree, fixed.

dixudx commented on keps/sig-node/20191025-image-pull-progress.md:
````diff
@@ -0,0 +1,402 @@
+container image will now look like:
+
+```
+> kubectl get pods -o events
````
we do have `kubectl get events --field-selector involvedObject.kind=Pod,involvedObject.name=pod1` for now.

dixudx replied:
And `kubectl describe` could also show related events.

saschagrunert replied:
Do you suggest in not modifying the command line at all? :)

saschagrunert replied:
The problem could be solved via the CRD, too.

zhangxiaoyu-zidif replied:
Personally speaking, it's better to show events by specified command. `-o events` is fine to me.

dixudx commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+- **Can the feature be disabled once it has been enabled (i.e., can we roll
+  back the enablement)?**
+
+  Yes, but not during a image pull (already established gRPC stream).
```
> but not during a image pull (already established gRPC stream)

Since kubelet got the `ctx`, it could still close the `ctx` when the feature got disabled.

saschagrunert replied:
I agree, I will make that more clear.

saschagrunert commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  Via the kubelet dynamic configuration feature, where we will add a dedicated
+  option for it.
```
Okay, I make that more clear.

saschagrunert commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,402 @@
+  image at the same time.
+
+  In any configuration case we would receive a starting and ending event for
+  the image pull, which would result in at least 200 QPS to the API Server.
```
Okay, I removed it since the event will be thrown in any case.

saschagrunert commented:
Thanks for the review @dixudx and @wojtek-t :pray:, I see two bigger open discussion points (linked in the PR body), where I would need further input (maybe from the SIGs as well).

saschagrunert commented:
I added a new commit on top which is now focused on the CRD approach.

johnbelamaric commented:
A CR per node means that when enabled, this is getting updated all the time. 99.9% of the time or more in production clusters there is no human watching for an image to be pulled. I imagine the primary use case here is in dev environments. Is that accurate? This seems like more of an ecosystem project than core K8s.

One of the big issues with this is all the chatter it creates for rarely used data.  Here's another option: 

What about a CR that is associated with a specific image; when you create a pod where you care about this sort of thing you can create that CR to track that you want to record pull progress (or in a dev environment use a webhook for that). A controller would watch pods and when it finds one using that image it would track the pull of that image on the node on which the pod is scheduled, and write back to the CR with the node name and the progress.

mattjmcnaughton commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,385 @@
+### Implementation Details/Notes/Constraints
+
+The major caveat for this implementation approach is the increased amount of
+resource updates from the kubelet during an image pull. This should be limited in a
```
To preface, I'm far from a networking expert so I apologize if I'm misunderstanding. But at a high level, it feels a little strange that (in my opinion) image pull progress is most useful when the image is taking a long time to download (i.e. there's some form of network congestion), but the steps we are taking to show image pull progress could actually make the congestion worse, because the kubelet is doing more communication with the API server?

In other words, I'm not sure if its worth having the image pull take longer so that we can see its progress.

mattjmcnaughton commented on keps/sig-node/20191025-image-pull-progress.md:
```diff
@@ -0,0 +1,385 @@
+
+During Kubernetes workload creation, the time needed to pull the container image
+from the remote location varies depending on multiple factors. In low-bandwidth
+scenarios the image pull step might consume a significant amount time
```
I'd love to learn more about the scenarios in which this feature would be most useful and how often they occur. When you get a chance, would you mind commenting on the following:

- Is this request one we've heard from users of k8s?
- How often do operators actively monitor a container image being pulled from a remote location?
- What would happen if we didn't undertake work on this feature? I.e. what are the risks to k8s and its applications?

mattjmcnaughton replied:
Actually, I answered my own questions reading https://github.com/kubernetes/kubernetes/issues/19077 :)

Sounds like something the community is interested in!

If there's additional thought you have on my questions, feel free to share.

justaugustus commented:
/remove-lifecycle rotten

ShadowJonathan commented:
@justaugustus

dims commented:
/assign @derekwaynecarr @dchen1107 

Derek, Dawn, Any interest in this work?

saschagrunert commented:
Since I've got some back pressure from the community I'm still wondering if we want to abandon this KEP and close the corresponding issue in k/k to clearly state that we do not want this.

OTOH I'm also happy to continue my path here. :shrug:

StoyanStAtanasov commented:
/reopen
While waiting for 6GB image to be pulled, not knowing what the status is, I found this PR and the associated issue.
@saschagrunert Let's not abandon this. I think we need this as there are a lot of images that can be huge in the AI area.

saschagrunert commented:
> While waiting for 6GB image to be pulled, not knowing what the status is, I found this PR and the associated issue. @saschagrunert Let's not abandon this. I think we need this as there are a lot of images that can be huge in the AI area.

In the meanwhile, we implemented pull progress metrics to CRI-O: https://github.com/cri-o/cri-o/blob/main/tutorials/metrics.md#available-metrics

@StoyanStAtanasov, we need a larger group within the k/k community to advocate this, otherwise I don't see a reasonable path forward with this enhancement.
